### PR TITLE
[ISSUE-3712] Separate message preparation part for SendMessageInterceptor

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
issue [#3712](https://github.com/GetStream/stream-chat-android/issues/3712)

### 🎯 Goal

Create a interceptor only for message preparation before message can be sent. This is a part of ISSUE-3712, but it doesn't solve it completely. 

### 🛠 Implementation details

- Creates a new interceptor and uses it instead of preparing messages inside the `SendMessageInterceptorImpl`. 

### 🧪 Testing

Send attachments and check that they work as before.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- ~[ ] Comparison screenshots added for visual changes~
- ~[ ] Affected documentation updated (KDocs, docusaurus, tutorial)~

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 1]
```diff
@@ -1309,12 +1310,19 @@ internal constructor(
...
     ): Call<Message> {
         val relevantPlugins = plugins.filterIsInstance<SendMessageListener>().also(::logPlugins)
-        val relevantInterceptors = interceptors.filterIsInstance<SendMessageInterceptor>()
-        return CoroutineCall(scope) {
+        val sendMessageInterceptors = interceptors.filterIsInstance<SendMessageInterceptor>()
+        val prepareMessageInterceptor = interceptors.find { interceptor ->
+            interceptor is PrepareMessageInterceptor
+        } as? PrepareMessageInterceptor
+
+        val preparedMessage = getCurrentUser()?.let { user ->
```
Comment (JcMinarro) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:1318]: `getCurrentuser()` will return `null` when WS is not connected.
There is another method that obtains a user from the userCredentials (`ChatClient::getStoredUser`)
As far as we want to be able to send messages while WS is not connected, we should try to use it

**Review by kanat — COMMENTED** [review]
These changes are good in general, however, we probably should avoid changing the public API:
- we made the process of intercepting message sending a bit complex by introducing an additional layer of interception
- this change may potentially break customer who passes their own `SendMessageIntereceptor` because we pass an altered `message` by `PrepareMessageInterceptor`  instead of an original one.

I wonder if we can keep running with a single `SendMessageInterceptor`, which is quite simple for understanding, and make this separation internally. For instance, we could extract both parts from `SendMessageInterceptorImpl` and leave inside `SendMessageInterceptorImpl` 2 delegates, one for preparation and the second one for uploading.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package io.getstream.chat.android.client.interceptor.message
+
+import android.content.Context
+import android.net.ConnectivityManager
+import io.getstream.chat.android.client.interceptor.message.internal.PrepareMessageInterceptorImpl
+import io.getstream.chat.android.client.network.NetworkStateProvider
+
+public class PrepareMessageInterceptorFactory {
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt:24]: Should we mark this with `@InternalStreamChatAPI`?
Also let's add KDocs

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    public fun create(context: Context): PrepareMessageInterceptor {
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt:26]: And KDocs here as well

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        val networkStateProvider =
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt:27]: We should replace it with `ChatClientState` after merging #3795, right?

File: stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/interceptor/internal/SendMessageInterceptorImpl.kt [hunk 5]
```diff
@@ -69,78 +63,33 @@ internal class SendMessageInterceptorImpl(
...
     ): Result<Message> {
         val channel = logic.channel(channelType, channelId)
+
         message.populateMentions(channel.toChannel())
 
+        // Update flow in channel controller
+        channel.upsertMessage(message)
+        // TODO: an event broadcasting feature for LOCAL/offline events on the LLC would be a cleaner approach
+        // Update flow for currently running queries
+        logic.getActiveQueryChannelsLogic().forEach { query -> query.refreshChannel(channel.cid) }
```
Comment (adasiewiczr) [inline stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/interceptor/internal/SendMessageInterceptorImpl.kt:73]: We should apply changes from #3820

Comment (leandroBorgesFerreira) [reply stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/interceptor/internal/SendMessageInterceptorImpl.kt]: The link sends me back to this PR

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt (hunk 4 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/PrepareMessageInterceptorFactory.kt]: ok

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/internal/PrepareMessageLogicImpl.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.network.NetworkStateProvider
+import io.getstream.chat.android.client.utils.SyncStatus
+import io.getstream.chat.android.client.utils.internal.getMessageType
+import java.util.Date
+import java.util.UUID
+
+internal class PrepareMessageLogicImpl(
+    private val networkStateProvider: NetworkStateProvider,
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/interceptor/message/internal/PrepareMessageLogicImpl.kt:32]: Can we use`ClientState` instead? 🤔

File: stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/interceptor/internal/SendMessageInterceptorImpl.kt (hunk 5 above)
Comment (adasiewiczr) [reply stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/interceptor/internal/SendMessageInterceptorImpl.kt:83]: My bad, it should be #3933

**Review by kanat — APPROVED** [review]
LGTM

File: stream-chat-android-client/src/test/java/io/getstream/chat/android/client/interceptor/message/internal/PrepareMessageLogicImplTest.kt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        )
+
+        val preparedMessage = prepareMessageInterceptorImpl.prepareMessage(
+            messageWithAttachments,
+            randomString(),
+            randomString(),
+            randomUser()
+        )
+
+        preparedMessage.syncStatus `should be equal to` SyncStatus.AWAITING_ATTACHMENTS
```
Comment (JcMinarro) [inline stream-chat-android-client/src/test/java/io/getstream/chat/android/client/interceptor/message/internal/PrepareMessageLogicImplTest.kt:54]: Can we verify the rest of fileds that are updated inside of `prepareMessage()`?

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**